### DISCUSSION TITLE
node fails at the end

### DISCUSSION THREAD

**Discussion by fotis93** [gstavrinos/ros2_kart_racing#24] [Q&A] [answered]
Καλησπερα ξανα επιχείρησα να φτιαξω το συγκεκριμένο node  δεν ειναι lifecycle αλλα μου σκαει στο τελος 
[server node.odt](https://github.com/gstavrinos/ros2_kart_racing/files/11057353/server.node.odt)

**Comment by gstavrinos** [comment, marked as answer]
Καλημέρα, δες λίγο τί κάνεις μέσα στο while. Όση περισσότερη ώρα τρέχει αυτό το while, τόσο πιο "επίπονο" θα είναι για το μηχάνημα! Επίσης, δες πώς χειρίζεται τα inputs από το request του action. 

(Hint: Μπορείς να βάζεις κώδικα και απευθείας στα discussions, χρησιμοποιώντας τα code tags)

**Comment by fotis93** [comment]
`import time 

import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node

from myaction.action import Myaction
from geometry_msgs.msg import Twist


class myaction(Node):

    def __init__(self):
        super().__init__('my_action_server')

        #published topic
        self.declare_parameter('Putopic','published_velocity')   
        #action server topic
        self.declare_parameter('Actopic','myaction1')   
        #publication rate
        self.declare_parameter('Purate', 0.5) 
        d1=self.get_parameter('Putopic').get_parameter_value().string_value
        d2=self.get_parameter('Actopic').get_parameter_value().string_value
        d3=self.get_parameter('Purate').get_parameter_value().integer_value 



        self._action_server = ActionServer(
            self,
            Myaction,
            d2,
            self.execute_callback)
        self.timer_period=d3
        self.publisher_ = self.create_publisher(Twist, d1, 10)

    def execute_callback(self, goal_handle):
        start=time.time()

        self.get_logger().info('Executing goal...')
        u=0#inital speed
        acc=goal_handle.request.linear_acceleration_x
        feedback_msg=Myaction.Feedback()

        max=goal_handle.request.max_linear_velocity_x# randomly chosen 

        vel=[]
        totaltime=0
        while(u<=max):
            end = time.time()
            tim=(end-start)
            u=u+acc*tim
            start=time.time()

            feedback_msg.current_linear_velocity_x=u
            feedback_msg.current_elapsed_time=tim

            msg=Twist()
            msg.linear.x=u
            self.publisher_.publish(msg)


            goal_handle.publish_feedback(feedback_msg)
            vel.append(u)#creating a list of instant velocities
            totaltime+=tim#summing to get the total time elapsed 

        goal_handle.succeed()
        result=Myaction.Result()
        result.total_elapsed_time=totaltime
        result.linear_velocities_x=vel
        return result
    


def main(args=None):
    rclpy.init(args=args)

    my_action_server = myaction()

    rclpy.spin(my_action_server)
   # my_action_server.destroy_node()
    #rclpy.shutdown()


if __name__ == '__main__':
    main()
`

**Comment by fotis93** [comment]
Αν βαλω το publish μεσα στο execute_callback δεν εχω θεμα και μου τρεχει. Χανω ομως το publishing rate  εκτος αν το κανω χειροκινητα. Σε αυτη τη περιπτωση μηπως θελουμε και τη ταχυτητα να την υπολογιζουμε ανα καποια δευτερολεπτα και οχι οσο πιο συχνα μπορω με το while?

**Reply by gstavrinos** [reply]
Τώρα δουλεύει γιατί έχεις αφαιρέσει τελείως τον publisher. Δοκίμασε να τον βάλει σε άλλο σημείο, έτσι ώστε να μην χάσεις και τη δυνατότητα του rate.
